Source/Core/Core/IOS/Network/KD/NetKDRequest.cpp

```diff
@@ -400,9 +408,103 @@ NWC24::ErrorCode NetKDRequestDevice::KDCheckMail(u32* mail_flag, u32* interval)
   return NWC24::WC24_OK;
 }
 
+NWC24::ErrorCode NetKDRequestDevice::DetermineDownloadTask(u16* entry_index,
+                                                           std::optional<u8>* subtask_id) const
+{
+  // As the scheduler does not tell us which entry to download, we must determine that.
+  // A correct entry is one that hasn't been downloaded the longest compared to other entries.
+  // We first need current UTC.
+  auto time_device =
+      std::static_pointer_cast<NetKDTimeDevice>(GetIOS()->GetDeviceByName("/dev/net/kd/time"));
+  const u32 current_utc = time_device->GetAdjustedUTC();
+  u32 lowest_timestamp = std::numeric_limits<u32>::max();
+
+  for (u16 i = 0; i < static_cast<u16>(NWC24::NWC24Dl::MAX_ENTRIES); i++)
+  {
+    if (!m_dl_list.DoesEntryExist(i))
+      continue;
+
+    if (m_dl_list.SkipSchedulerDownload(i))
+      continue;
+
+    const u32 next_dl_time = m_dl_list.GetNextDownloadTime(i);
+
+    // First determine if UTC is greater than the next download time.
+    if (current_utc < next_dl_time)
+      continue;
+
+    // If this task's next download time is less than the lowest_timestamp, this is the task we
+    // want. However, we must determine if this has a subtask and wants to be downloaded.
+    if (next_dl_time < lowest_timestamp)
+    {
+      if (m_dl_list.HasSubtask(i))
+      {
+        // No valid subtask found or downloading is disabled.
+        NWC24::ErrorCode code = DetermineSubtask(i, subtask_id);
+        if (code != NWC24::WC24_OK)
+          continue;
+      }
+
+      lowest_timestamp = next_dl_time;
```

Comment:
If the index does not have a subtask, should we be updating the `lowest_timestamp`?

Reply:
I should have placed the comment on line 442 in the failure path, but subtasks are not a necessity for tasks. The only channel I know of that uses subtasks is the News Channel.